pkg/scheduler/plugins/gang/gang_plugin.go

```diff
@@ -0,0 +1,507 @@
+			if GetNamespaceSplicingName(pod.Namespace,
+				GetGangNameByPod(pod)) == gangId {
+				pod = podOriginal.DeepCopy()
+				if pod.Annotations == nil {
+					pod.Annotations = map[string]string{}
+				}
+				pod.Annotations[extension.AnnotationGangTimeout] = "true"
+				patchBytes, err := generatePodPatch(podOriginal, pod)
+				if err != nil {
+					klog.Errorf("unReserve generatePodPatch got err:%v", err.Error())
+					continue
+				}
+				if string(patchBytes) == "{}" {
+					continue
+				}
+				err = retry.OnError(
+					retry.DefaultRetry,
+					errors.IsTooManyRequests,
+					func() error {
+						_, err = p.frameworkHandler.ClientSet().CoreV1().Pods(pod.Namespace).Patch(ctx, pod.Name,
+							types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
+						if err != nil {
+							klog.Errorf("unReserve when patch gang timeout annotation to pod:%v, err:%v",
+								GetNamespaceSplicingName(pod.Namespace, pod.Name), err.Error())
+						}
+						return err
+					})
+				if err != nil {
+					klog.Errorf("Failed to patch gang timeout annotation after retry to pod:%v, err:%v",
+						GetNamespaceSplicingName(pod.Namespace, pod.Name), err.Error())
+				}
+			}
+		}
+
+		// release resource of all assumed children of the gang
+		p.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+			if GetNamespaceSplicingName(waitingPod.GetPod().Namespace,
+				GetGangNameByPod(waitingPod.GetPod())) == gangId {
+				klog.Errorf("unReserve rejects the pod name:%v from Gang %s due to timeout",
+					GetNamespaceSplicingName(pod.Namespace, pod.Name), gangId)
+				waitingPod.Reject(p.Name(), "optimistic rejection in unReserve due to timeout")
+			}
+		})
+	}
+}
+
+// PostBind just update the gang's BoundChildren
+func (p *GangPlugin) PostBind(ctx context.Context, _ *framework.CycleState, pod *corev1.Pod, nodeName string) {
+	gangId := GetNamespaceSplicingName(pod.Namespace,
+		GetGangNameByPod(pod))
+	gang := p.gangCache.GetGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			GetNamespaceSplicingName(pod.Namespace, pod.Name))
+		return
+	}
+
+	gang.AddBoundPod(pod)
+}
+
+func (p *GangPlugin) RecoverGangCache() {
+	// recover podGroup First
+	pgLister := p.pgLister
+	pgList, err := pgLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podGroupList List error %v", err.Error())
+	}
+	for _, podGroup := range pgList {
+		p.gangCache.OnGangAdd(podGroup)
+
+	}
+	//recover gang from pod
+	podLister := p.podLister
+	podsList, err := podLister.List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("RecoverGangCache podsList List error %v", err.Error())
+	}
+	for _, pod := range podsList {
+		p.gangCache.OnPodAdd(pod)
+
+		if pod.Spec.NodeName != "" {
+			gangId := GetNamespaceSplicingName(pod.Namespace,
+				GetGangNameByPod(pod))
+			gang := p.gangCache.GetGangFromCache(gangId)
+			gang.AddAssumedPod(pod)
```

Comment:
AddAssumedPod is error; shoud like AddBoundPod

Reply:
done